### PR TITLE
Support nested routes for has many through

### PR DESCRIPTION
**Pull request by chrisandreae**
Nested routes previously didn't, now they do.

### CODE REVIEW

File: lib/view_model/active_record/association_manipulation.rb [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+      else
+        updated_viewmodels.first
+      end
+    end
+  end
+
+  # Removes the association between the models represented by this viewmodel and
+  # the provided associated viewmodel. The associated model will be
+  # garbage-collected if the assocation is specified with `dependent: :destroy`
+  # or `:delete_all`
```
Comment (thefloweringash) [inline lib/view_model/active_record/association_manipulation.rb]: Is this comment still true? It looks like this always calls `vm.destroy!`.

File: lib/view_model/active_record/association_manipulation.rb [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+      else
+        direct_viewmodel_class = association_data.viewmodel_class
+        root_update_data, referenced_update_data = construct_direct_append_updates(association_data, subtree_hashes, references)
+      end
+
+      update_context = UpdateContext.build!(root_update_data, referenced_update_data, root_type: direct_viewmodel_class)
+
+      # Provide information about what was updated
+      deserialize_context.updated_associations = root_update_data
+                                                   .map { |upd| upd.updated_associations }
```
Comment (chrisandreae) [inline lib/view_model/active_record/association_manipulation.rb]: yes

File: lib/view_model/active_record/association_manipulation.rb [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                                                   .inject({}) { |acc, assocs| acc.deep_merge(assocs) }
+
+      # Set new parent
+      new_parent = ViewModel::ActiveRecord::UpdateOperation::ParentData.new(direct_reflection.inverse_of, self)
+      update_context.root_updates.each { |update| update.reparent_to = new_parent }
```
Comment (chrisandreae) [inline lib/view_model/active_record/association_manipulation.rb]: I believe it does. Darn.

File: lib/view_model/active_record/association_manipulation.rb [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+      association_scope = association_scope.order(direct_viewmodel._list_attribute_name)
+    end
+
+    if association_data.through?
+      association_scope = associated_viewmodel.model_class
+                            .joins(association_data.indirect_reflection.inverse_of.name)
+                            .merge(association_scope)
+    end
+
+    association_scope = association_scope.merge(scope) if scope
```
Comment (chrisandreae) [inline lib/view_model/active_record/association_manipulation.rb]:
```> direct_association_scope = Category.first.association(:category_courses).association_scope; direct_association_scope.to_sql
 => "SELECT \"category_courses\".* FROM \"category_courses\" WHERE \"category_courses\".\"category_id\" = '9eee229a-d32c-11e6-92f4-ff2111ec3da8'"
2.3.2 :010 > Course.joins(:category_courses).merge(direct_association_scope).to_sql
 => "SELECT \"courses\".* FROM \"courses\" INNER JOIN \"category_courses\" ON \"category_courses\".\"course_id\" = \"courses\".\"id\" WHERE \"category_courses\".\"category_id\" = '9eee229a-d32c-11e6-92f4-ff2111ec3da8'"```

File: lib/view_model/active_record/association_manipulation.rb (hunk 1 above)
Comment (chrisandreae) [inline lib/view_model/active_record/association_manipulation.rb]: No, this is now just wrong. Works for `through:`, but no longer for direct associations.

File: lib/view_model/active_record/association_manipulation.rb [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+      UpdateData.new(direct_viewmodel_class, existing_id, existing_id.nil?,
+                     { indirect_reflection.name.to_s => { ViewModel::REFERENCE_ATTRIBUTE => ref_name }},
+                     [ref_name] )
+    end
+
+    return direct_update_data, referenced_update_data
+  end
+
+  def convert_updates_to_references(indirect_update_data, referenced_update_data)
```
Comment (chrisandreae) [inline lib/view_model/active_record/association_manipulation.rb]: fixed

File: lib/view_model/active_record/association_manipulation.rb (hunk 3 above)
Comment (chrisandreae) [inline lib/view_model/active_record/association_manipulation.rb]: fixed

File: lib/view_model/active_record/collection_nested_controller.rb [hunk 6]
```diff
@@ -56,9 +66,27 @@ def append(serialize_context: new_serialize_context, deserialize_context: new_de
...
+
+  private
+
+  def parse_relative_position(name)
+    id = parse_uuid_param(name, default: nil)
+
+    if id
+      if association_data.polymorphic?
+        type_name = parse_param("#{name}_type")
+        type = ViewModel::Registry.for_view_name(type_name)
```
Comment (chrisandreae) [inline lib/view_model/active_record/collection_nested_controller.rb]: fixed

File: test/unit/view_model/active_record/has_many_test.rb [hunk 7]
```diff
@@ -228,6 +276,90 @@ def test_edit_has_many
     assert(Child.where(id: c1.id).blank?)
   end
 
+  def test_append_associated_move_has_many
```
Comment (chrisandreae) [inline test/unit/view_model/active_record/has_many_test.rb]: now also includes test of ownership change

**Pull request merged**